tests/Doctrine/Tests/ORM/Functional/Ticket/Issue6313Test.php

```diff
@@ -0,0 +1,245 @@
+<?php
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\StringType;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\ORM\AbstractQuery;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use Doctrine\Common\Collections\ArrayCollection;
+
+final class Issue6313Test extends OrmFunctionalTestCase
+{
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->_schemaTool->createSchema(
+            [
+                $this->_em->getClassMetadata(Issue6313_Unit::class),
+                $this->_em->getClassMetadata(Issue6313_Port::class),
+                $this->_em->getClassMetadata(Issue6313_Container::class),
+                $this->_em->getClassMetadata(Issue6313_Container1::class),
+                $this->_em->getClassMetadata(Issue6313_Container2::class),
+                $this->_em->getClassMetadata(Issue6313_Container3::class),
+                $this->_em->getClassMetadata(Issue6313_Container4::class),
+            ]
+        );
+    }
+
+    public function testOrphanWithCollectionShouldBeDeletedOnOneToManyRelation()
+    {
+        $unit1 = new Issue6313_Unit1();
+
+        $port1 = new Issue6313_Port();
+        $port1->addContainer(new Issue6313_Container1());
+        $port1->addContainer(new Issue6313_Container2());
+
+        $port2 = new Issue6313_Port();
+        $port2->addContainer(new Issue6313_Container1());
+        $port2->addContainer(new Issue6313_Container2());
+
+        $unit1->addPort($port1);
+        $unit1->addPort($port2);
+
+        $this->_em->persist($unit1);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $unit1 = $this->_em->find(Issue6313_Unit::class, $unit1->id);
+        $port1 = $this->_em->find(Issue6313_Port::class, $port1->id);
+        $unit1->removePort($port1);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $unit1 = $this->_em->find(Issue6313_Unit::class, $unit1->id);
+
+        self::assertEquals(1, $unit1->ports->count());
+    }
+}
+
+/**
+ * @Entity
+ * @InheritanceType("SINGLE_TABLE")
+ * @DiscriminatorColumn(name="type_id", type="integer")
+ * @DiscriminatorMap({1 = "Issue6313_Unit1", 2 = "Issue6313_Unit2"})
+ */
+abstract class Issue6313_Unit
```

Comment:
It's part of my app and I need this logic. You are right, test need to be reduced for orphanRemoval, but when I test this code I found one more bug, so I open new issue 6318. #6318 different from this because its show wrong CommitOrderCalculator for One-To-Many self-ref with JOINED table.